### PR TITLE
Stopwords functions cf

### PR DESCRIPTION
**Pull request by cforgaci**
I made some cleanup in the stopwords removal functions. I add some comments on the files where I think we still have unresolved issues.

### CODE REVIEW

File: R/clean.R [hunk 1]
```diff
@@ -11,7 +11,7 @@ clean_basic <- function(text) {
   out_text <- text |>
     tolower() |>
     str_replace_all("\\s{1,}", " ") |>
-    str_remove_all("[^0-9a-z ]") |>
+    str_remove_all("[[:punct:]]") |>
```
Comment (cforgaci) [inline R/clean.R:14]: I added this change to avoid that non-English characters are removed. We have many proper names like that.

File: R/clean.R [hunk 2]
```diff
@@ -95,10 +95,13 @@ find_meta_stopwords <- function(metadata,
 #' @param add_stopwords A vector of words to be added to the urbanism stopwords
 #' @return A vector of urbanism-related stopwords
 #' @export
-urbanism_stopwords <- function(add_stopwords = NULL) {
-  stop_words <-
-    c(add_stopwords, "city", "urban", "urbanism", "hab", "km")
+subject_stopwords <- function(add_stopwords = NULL, convert_to_regex = TRUE) {
+  stop_words <- c(add_stopwords, "city", "cities", "urban", "urbanism", "hab", "km",
+                  "emu", "european postgraduate master in urbanism", "advisor",
+                  "prof")
```
Comment (cforgaci) [inline R/clean.R:101]: I added a few more subject-specific words here.

File: R/clean.R [hunk 3]
```diff
@@ -95,10 +95,13 @@ find_meta_stopwords <- function(metadata,
 #' @param add_stopwords A vector of words to be added to the urbanism stopwords
 #' @return A vector of urbanism-related stopwords
 #' @export
-urbanism_stopwords <- function(add_stopwords = NULL) {
-  stop_words <-
-    c(add_stopwords, "city", "urban", "urbanism", "hab", "km")
+subject_stopwords <- function(add_stopwords = NULL, convert_to_regex = TRUE) {
```
Comment (cforgaci) [inline R/clean.R:98]: I gave this function a more generic name to reflect what in literature is called subject-specific stopwords. Now, with this change, I am wondering if we should write a wrapper function urbanism_stopwords or if we should pass the vector of stopwords as an argument in the paper when calling `subject_stopwords()`. What do you think?

File: analysis/paper/paper.qmd [hunk 4]
```diff
@@ -101,29 +101,8 @@ emu_theses$text <- convert_pdf_text(pdf_paths)
...
-# Preview theses
-emu_theses %>% 
-  arrange(graduation_year) %>% 
-  pull(text_clean) %>% 
-  str_sub(1, 500)
-
-# Preview theses
-hist(str_length(emu_theses$text_clean))
-
-Encoding(emu_theses$text_clean)
```
Comment (cforgaci) [inline analysis/paper/paper.qmd:126]: I removed some preview code that was not necessary for the paper.

File: analysis/paper/paper.qmd [hunk 5]
```diff
@@ -203,20 +182,15 @@ Word-level tokenisation was applied with the `unnest_tokens()` function of the `
...
-  unnest_tokens(output = word, input = title) %>%   ### add words from the title
-  select(word) %>% 
-  rbind(., data.frame(word = c("preface", "foreword", "introduction", "conclusion", "thesis", 
-                               "source", "author"))) %>% 
-  # rbind(., data.frame(word = c("city", "urban", "urbanism"))) %>% 
-  rbind(., data.frame(word = c("hab", "km")))
+emu_theses$text_clean <- emu_theses$text_clean |>
+  str_remove_all(find_meta_stopwords(emu_theses)) |> 
+  str_remove_all(subject_stopwords()) |> 
+  str_remove_all(document_stopwords())
```
Comment (cforgaci) [inline analysis/paper/paper.qmd:188]: I moved these lines from the `clean-text` code chunk to the `stop-words` code chunk. 

This removes:
1. all metadata stopwords on the document level and
2. all subject-specific and document stopwords on the level of the corpus

These are stop phrases, to be precise, so it is better to remove them before tokenizing the text.

File: R/clean.R (hunk 1 above)
Comment (alwil) [reply R/clean.R:14]: great!

File: R/clean.R (hunk 2 above)
Comment (alwil) [reply R/clean.R:101]: this is very specific: "european postgraduate master in urbanism". Maybe we want to keep it out of the function and add it in the argument `add_stopwords` , when calling it?

Comment (cforgaci) [reply R/clean.R:101]: Yes, I would even suggest we move out all subject-specific words to the paper, also to make them explicit to the reader. In this corpus, "emu" and "european postgraduate master of urbanism" might be as frequent as more general terms such as "city" and "urban".

**Review by alwil — COMMENTED** [review]
I would suggest keeping `urbanism_stopwords()` and let's discuss adding specific terms inside the function.

**Review by alwil — APPROVED** [review]

**Pull request merged**